File: Packages/WhoopStore/Sources/WhoopStore/BackupSettings.swift
import Foundation

/// The `settings.json` payload inside a `.noopbak` backup (#1000).
///
/// A `.noopbak` is a ZIP whose first entry is the SQLite database. That round-trips every row, but the
/// user's profile (age / sex / weight / height / HR-max override) and display preferences live in
/// UserDefaults (SharedPreferences on Android), so a restore onto a fresh device silently reset them —
/// the "restore doesn't bring back settings/weight/height" half of #1000. This adds a SECOND, optional
/// ZIP entry — `settings.json`, a flat JSON object — carrying exactly one WHITELISTED set of keys.
///
/// The whitelist is the contract, defined once per platform and mirrored byte-for-byte by Android's
/// `BackupSettingsCodec` (same canonical key strings, same JSON kinds). Only stable, user-set,
/// non-device-specific values are allowed. NEVER add device ids, peripheral ids, tokens, sync cursors,
/// or anything anonymity-sensitive: backups get copied into cloud folders and attached to GitHub
/// issues, so this file must stay safe to share. Unknown keys in an incoming `settings.json` are
/// dropped on the floor; a backup with no `settings.json` (every pre-#1000 backup) is simply a
/// DB-only restore, exactly as before.
///
/// Pure JSON + UserDefaults mapping only — no ZIP code here. The container work stays in the app's
/// `DataBackup`; this lives in the package so the codec and the defaults round-trip are unit-testable
/// headlessly (`swift test --filter BackupSettingsTests`).
public enum BackupSettings {

    /// Canonical entry name inside the `.noopbak` ZIP. Matches the Android exporter byte-for-byte.
    public static let entryName = "settings.json"

    /// The JSON kind a whitelisted key must decode to. Anything else (wrong type, JSON bool posing
    /// as a number, nested objects) is dropped rather than guessed at.
    public enum Kind: Sendable {
        case int, double, string
    }

    /// THE whitelist — the only keys `settings.json` may carry, keyed by their CANONICAL
    /// (platform-neutral) names. Mirrored exactly by Android's `BackupSettingsCodec.WHITELIST`.
    ///
    /// Profile: the body metrics that power HR zones / calories / recovery baselines, plus the manual
    /// HR-max override (`profile.hrMax`, 0 = auto/Tanaka). Display: the metric/imperial system, the
    /// separate temperature override ("" = match the system), and the Effort axis (#268) — the three
    /// display prefs that exist with identical semantics on both platforms. Deliberately EXCLUDED:
    /// step calibration (per-strap, not per-person), the avatar blob (bulky, and not "settings"),
    /// steps-engine fitted outputs (derived), and every noop.* toggle that is device- or
    /// install-specific.
    public static let whitelist: [String: Kind] = [
        "profile.age": .int,
        "profile.sex": .string,
        "profile.weightKg": .double,
        "profile.heightCm": .double,
        "profile.waistCm": .double,
        "profile.hrMax": .int,
        "units.system": .string,
        "units.temperature": .string,
        "effort.scale": .string,
    ]

    /// Canonical JSON key → this platform's UserDefaults key. Identity everywhere except
    /// `profile.hrMax`, which UserDefaults stores as `profile.hrMaxOverride` (see `ProfileStore.K`).
    /// Android maps the same canonical keys onto its own SharedPreferences names.
    public static let appleDefaultsKey: [String: String] = [
        "profile.age": "profile.age",
        "profile.sex": "profile.sex",
        "profile.weightKg": "profile.weightKg",
        "profile.heightCm": "profile.heightCm",
        "profile.waistCm": "profile.waistCm",
        "profile.hrMax": "profile.hrMaxOverride",
        "units.system": "units.system",
        "units.temperature": "units.temperature",
        "effort.scale": "effort.scale",
    ]

    // MARK: - Snapshot / apply (UserDefaults boundary)

    /// The whitelisted values currently SET in `defaults`, keyed canonically — the export-side
    /// snapshot. Keys the user never touched are omitted (not defaulted), so restoring this backup
    /// on another device only overwrites what was genuinely set here and leaves the rest of the
    /// target's settings alone.
    public static func snapshot(from defaults: UserDefaults) -> [String: Any] {
        var out: [String: Any] = [:]
        for (canonical, kind) in whitelist {
            guard let storageKey = appleDefaultsKey[canonical],
                  let raw = defaults.object(forKey: storageKey),
                  let coerced = coerce(raw, to: kind) else { continue }
            out[canonical] = coerced
        }
        return out
    }

    /// Write the whitelisted `values` (canonical keys) into `defaults` under this platform's storage
    /// keys — the restore-side apply. Non-whitelisted keys and wrong-typed values are ignored. The
    /// caller decides WHEN (DataBackup applies only after a successful DB swap, never on a failed or
    /// rolled-back restore).
    public static func apply(_ values: [String: Any], to defaults: UserDefaults) {
        for (canonical, kind) in whitelist {
            guard let raw = values[canonical],
                  let coerced = coerce(raw, to: kind),
                  let storageKey = appleDefaultsKey[canonical] else { continue }
            defaults.set(coerced, forKey: storageKey)
        }
        // #146: the whitelist carries an Int `profile.age`, not a Date, so a restore can only bring an
        // age. Clear any stale `profile.dateOfBirth` on this device whenever an age is applied, so
        // `ProfileStore` re-derives the date of birth from the RESTORED age on next launch instead of a
        // pre-existing local DOB silently overriding the restore. (A restore forces a relaunch, so the
        // re-derivation always runs.) `dobDefaultsKey` is spelled out here rather than added to the
        // whitelist to keep the cross-platform JSON contract unchanged.
        if values["profile.age"] != nil {
            defaults.removeObject(forKey: dobDefaultsKey)
        }
    }

    /// UserDefaults key `ProfileStore` stores the #146 date of birth under. Not a whitelisted backup
    /// key (a Date can't ride the Int/Double/String JSON contract) — only cleared on restore so the
    /// restored Int age re-derives it. Must match `ProfileStore.K.dateOfBirth`.
    static let dobDefaultsKey = "profile.dateOfBirth"

    // MARK: - JSON codec

    /// Encode the whitelisted subset of `values` as the flat `settings.json` object. Returns nil when
    /// nothing whitelisted is present (the exporter then writes a DB-only backup — indistinguishable
    /// from a legacy one, which is exactly the right degrade). `.sortedKeys` keeps the output
    /// deterministic so identical settings always produce identical bytes.
    public static func encode(_ values: [String: Any]) -> Data? {
        var filtered: [String: Any] = [:]
        for (key, kind) in whitelist {
            guard let raw = values[key], let coerced = coerce(raw, to: kind) else { continue }
            filtered[key] = coerced
        }
        guard !filtered.isEmpty else { return nil }
        return try? JSONSerialization.data(withJSONObject: filtered, options: [.sortedKeys])
    }

    /// Decode a `settings.json` payload down to its whitelisted, correctly-typed subset. Malformed
    /// JSON, a non-object root, unknown keys, and wrong-typed values all degrade to "fewer keys" —
    /// never an error, because a bad settings entry must not fail a restore whose DB half is fine.
    public static func decode(_ data: Data) -> [String: Any] {
        guard let obj = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else { return [:] }
        var out: [String: Any] = [:]
        for (key, kind) in whitelist {
            guard let raw = obj[key], let coerced = coerce(raw, to: kind) else { continue }
            out[key] = coerced
        }
        return out
    }

    // MARK: - Coercion

    /// Coerce a JSON-decoded (or caller-supplied) value to the whitelist's declared kind, or nil if
    /// it can't represent one. JSON booleans arrive as `NSNumber` too, so they are explicitly refused
    /// for numeric kinds — `true` must never become age 1.
    private static func coerce(_ value: Any, to kind: Kind) -> Any? {
        switch kind {
        case .string:
            return value as? String
        case .int:
            guard let n = value as? NSNumber, !isBoolean(n) else { return nil }
            return n.intValue
        case .double:
            guard let n = value as? NSNumber, !isBoolean(n) else { return nil }
            return n.doubleValue
        }
    }

    private static func isBoolean(_ n: NSNumber) -> Bool {
        CFGetTypeID(n) == CFBooleanGetTypeID()
    }
}

File: android/app/src/main/java/com/noop/data/BackupSettings.kt
package com.noop.data

import android.content.Context
import com.noop.ui.HostedCardPrefs
import com.noop.ui.NoopPrefs
import com.noop.ui.ProfileStore
import com.noop.ui.UnitPrefs
import org.json.JSONObject

/**
 * The `settings.json` payload inside a `.noopbak` backup (#1000) — the Android twin of the Apple
 * `BackupSettings` in Packages/WhoopStore.
 *
 * A `.noopbak` is a ZIP whose first entry is the SQLite database. That round-trips every row, but the
 * user's profile (age / sex / weight / height / HR-max override) and display preferences live in
 * SharedPreferences (UserDefaults on Apple), so a restore onto a fresh device silently reset them —
 * the "restore doesn't bring back settings/weight/height" half of #1000. This adds a SECOND, optional
 * ZIP entry — `settings.json`, a flat JSON object — carrying exactly one WHITELISTED set of keys.
 *
 * The whitelist is the contract, defined once per platform and mirrored byte-for-byte by the Apple
 * `BackupSettings.whitelist` (same canonical key strings, same JSON kinds). Only stable, user-set,
 * non-device-specific values are allowed. NEVER add device ids, peripheral ids, tokens, sync cursors,
 * or anything anonymity-sensitive: backups get copied into cloud folders and attached to GitHub
 * issues, so this file must stay safe to share. Unknown keys in an incoming `settings.json` are
 * dropped; a backup with no `settings.json` (every pre-#1000 backup) is a DB-only restore, as before.
 *
 * [BackupSettingsCodec] is pure JSON + whitelist (plain-JVM unit-testable, no Context); the
 * SharedPreferences boundary lives in [BackupSettingsBridge] below.
 */
object BackupSettingsCodec {

    /** Canonical entry name inside the `.noopbak` ZIP. Matches the Apple exporter byte-for-byte. */
    const val ENTRY_NAME = "settings.json"

    /** The JSON kind a whitelisted key must decode to. Anything else is dropped, never guessed at. */
    enum class Kind { INT, DOUBLE, STRING }

    /**
     * THE whitelist — the only keys `settings.json` may carry, keyed by their CANONICAL
     * (platform-neutral) names. Mirrors the Apple `BackupSettings.whitelist` exactly.
     *
     * Profile: the body metrics that power HR zones / calories / recovery baselines, plus the manual
     * HR-max override (`profile.hrMax`, 0 = auto/Tanaka). Display: the metric/imperial system, the
     * separate temperature override ("" = match the system), and the Effort axis (#268). Deliberately
     * EXCLUDED: step calibration (per-strap, not per-person), the steps-engine fitted outputs
     * (derived), and every noop.* toggle that is device- or install-specific.
     */
    val WHITELIST: Map<String, Kind> = linkedMapOf(
        "profile.age" to Kind.INT,
        "profile.sex" to Kind.STRING,
        "profile.weightKg" to Kind.DOUBLE,
        "profile.heightCm" to Kind.DOUBLE,
        "profile.waistCm" to Kind.DOUBLE,
        "profile.hrMax" to Kind.INT,
        "profile.hrZoneThresholds" to Kind.STRING,
        "units.system" to Kind.STRING,
        "units.temperature" to Kind.STRING,
        "effort.scale" to Kind.STRING,
        // The ONE layout pref carried (#today-hosted-cards): the Trends/Sleep cards the user chose to host
        // in Today, a JSON [String] of ids. Unlike section order, this is a deliberate composition the user
        // built and expects across a restore. Its POSITION (the addedCards slot in today.sectionOrder) is
        // not carried, so on restore the set + internal order return but the section sits at its default.
        HostedCardPrefs.KEY_SELECTION to Kind.STRING,
        // #1361: the user's own custom journal BEHAVIOURS (a newline-joined list of names). The journal
        // EFFECTS ride the DB backup, but the behaviour DEFINITIONS live only in prefs, so a restore left
        // the entries referencing behaviours the logging catalog no longer offered. Platform-neutral key;
        // the same newline value is bridged out of the catalog on both platforms.
        // SCOPE: NAMES only — the wire carries no kind/group, so a numeric custom behaviour restores as a
        // plain .bool toggle (identical on both platforms; historical entries keep their DB numericValue).
        "journal.customBehaviors" to Kind.STRING,
    )

    /**
     * Encode the whitelisted subset of [values] as the flat `settings.json` object, or null when
     * nothing whitelisted is present (the exporter then writes a DB-only backup — indistinguishable
     * from a legacy one, which is exactly the right degrade).
     */
    fun encode(values: Map<String, Any?>): String? {
        val obj = JSONObject()
        for ((key, kind) in WHITELIST) {
            val coerced = coerce(values[key], kind) ?: continue
            obj.put(key, coerced)
        }
        return if (obj.length() == 0) null else obj.toString()
    }

    /**
     * Decode a `settings.json` payload down to its whitelisted, correctly-typed subset. Malformed
     * JSON, unknown keys and wrong-typed values all degrade to "fewer keys" — never an error, because
     * a bad settings entry must not fail a restore whose DB half is fine.
     */
    fun decode(json: String): Map<String, Any> {
        val obj = runCatching { JSONObject(json) }.getOrNull() ?: return emptyMap()
        val out = LinkedHashMap<String, Any>()
        for ((key, kind) in WHITELIST) {
            if (!obj.has(key)) continue
            coerce(obj.opt(key), kind)?.let { out[key] = it }
        }
        return out
    }

    /**
     * Coerce a JSON-decoded (or caller-supplied) value to the whitelist's declared kind, or null.
     * JSON booleans are not [Number]s on the JVM, so `true` can never become age 1 (the Apple side
     * refuses NSNumber-booleans explicitly for the same reason).
     */
    private fun coerce(value: Any?, kind: Kind): Any? = when (kind) {
        Kind.STRING -> value as? String
        Kind.INT -> (value as? Number)?.toInt()
        Kind.DOUBLE -> (value as? Number)?.toDouble()
    }
}

/**
 * The SharedPreferences boundary for [BackupSettingsCodec]: snapshot this device's whitelisted
 * settings for export, and re-apply a restored payload. Kept separate from the codec so the codec
 * stays plain-JVM testable (this object needs a real Context).
 *
 * Storage mapping (canonical key → where it actually lives here):
 *  - `profile.*`  → the `noop_profile` prefs via [ProfileStore.backupSnapshot]/[ProfileStore.applyBackup]
 *                   (canonical `profile.hrMax` ↔ ProfileStore's `hr_max_override`).
 *  - `units.*` / `effort.scale` → [NoopPrefs] under the SAME literal key strings as the canonical names
 *                   (they were already kept identical to the Apple @AppStorage keys).
 */
object BackupSettingsBridge {

    /** Journal-catalog storage keys in `noop_prefs` (#1361). The v2 items blob is the LIVE store; the
     *  legacy custom/hidden keys are read ONCE by `JournalCatalog.loadJournalCatalogItems` to migrate into
     *  v2, then never again. So export pulls custom names out of the v2 blob (parsed raw here to avoid a
     *  data→ui dependency — the JSON shape mirrors `JournalCatalog.encodeJournalCatalog`), and restore
     *  writes the names to the legacy custom key and CLEARS the v2 blob so the next load re-migrates them
     *  (a restore requires an app restart, #57). Must match `JournalCatalog`. Mirrors the iOS bridge. */
    private const val JOURNAL_CATALOG_V2_PREF = "noop.journalCatalogV2"
    private const val JOURNAL_LEGACY_CUSTOM_PREF = "noop.journalCustomQuestions"
    private const val JOURNAL_LEGACY_HIDDEN_PREF = "noop.journalHiddenQuestions"

    /** The whitelisted, user-SET settings of this device as the `settings.json` string, or null. */
    fun snapshotJson(context: Context): String? {
        val values = LinkedHashMap<String, Any>()
        values.putAll(ProfileStore.from(context).backupSnapshot())
        val noop = NoopPrefs.of(context)
        if (noop.contains(NoopPrefs.KEY_UNIT_SYSTEM)) {
            noop.getString(NoopPrefs.KEY_UNIT_SYSTEM, null)?.let { values["units.system"] = it }
        }
        if (noop.contains(NoopPrefs.KEY_TEMPERATURE_UNIT)) {
            noop.getString(NoopPrefs.KEY_TEMPERATURE_UNIT, null)?.let { values["units.temperature"] = it }
        }
        if (noop.contains(UnitPrefs.KEY_EFFORT_SCALE)) {
            noop.getString(UnitPrefs.KEY_EFFORT_SCALE, null)?.let { values["effort.scale"] = it }
        }
        if (noop.contains(HostedCardPrefs.KEY_SELECTION)) {
            noop.getString(HostedCardPrefs.KEY_SELECTION, null)?.let { values[HostedCardPrefs.KEY_SELECTION] = it }
        }
        // #1361: custom journal behaviours from the LIVE v2 catalog blob (the legacy key is read-once-
        // then-stale). Parse raw to avoid a data→ui dependency; shape mirrors JournalCatalog.
        noop.getString(JOURNAL_CATALOG_V2_PREF, null)?.let { blob ->
            val names = runCatching {
                val arr = org.json.JSONArray(blob)
                (0 until arr.length()).mapNotNull { i ->
                    arr.optJSONObject(i)?.takeIf { it.optBoolean("custom", false) }
                        ?.optString("canonical", "")?.takeIf { it.isNotEmpty() }
                }
            }.getOrDefault(emptyList())
            if (names.isNotEmpty()) values["journal.customBehaviors"] = names.joinToString("\n")
        }
        return BackupSettingsCodec.encode(values)
    }

    /**
     * Re-apply a restored `settings.json` to this device. The caller ([DataBackup.importFrom]) invokes
     * this only AFTER the DB swap succeeded — never on a failed or rolled-back restore. Keys absent
     * from the payload leave the device's current values alone; the profile setters clamp to their
     * normal ranges, so a hand-edited payload can't write absurd values.
     */
    fun apply(context: Context, json: String) {
        val values = BackupSettingsCodec.decode(json)
        if (values.isEmpty()) return

        ProfileStore.from(context).applyBackup(values)

        val editor = NoopPrefs.of(context).edit()
        (values["units.system"] as? String)?.let { editor.putString(NoopPrefs.KEY_UNIT_SYSTEM, it) }
        (values["units.temperature"] as? String)?.let { raw ->
            // "" is the Apple side's "match the length/mass system"; here that state is key-absent.
            if (raw.isEmpty()) editor.remove(NoopPrefs.KEY_TEMPERATURE_UNIT)
            else editor.putString(NoopPrefs.KEY_TEMPERATURE_UNIT, raw)
        }
        (values["effort.scale"] as? String)?.let { editor.putString(UnitPrefs.KEY_EFFORT_SCALE, it) }
        (values[HostedCardPrefs.KEY_SELECTION] as? String)?.let { editor.putString(HostedCardPrefs.KEY_SELECTION, it) }
        // #1361: restore custom behaviours — write the names to the legacy custom key, clear stale hidden,
        // and drop the v2 blob so the next catalog load re-migrates them (restart-gated, #57). Mirrors iOS.
        (values["journal.customBehaviors"] as? String)?.let { joined ->
            editor.putString(JOURNAL_LEGACY_CUSTOM_PREF, joined)
            editor.remove(JOURNAL_LEGACY_HIDDEN_PREF)
            editor.remove(JOURNAL_CATALOG_V2_PREF)
        }
        editor.apply()
    }
}
